pyopencl/array.py

```diff
@@ -621,9 +621,11 @@ def set(self, ary, queue=None, async_=None, **kwargs):
                     stacklevel=2)
 
         if self.size:
-            cl.enqueue_copy(queue or self.queue, self.base_data, ary,
+            event1 = cl.enqueue_copy(queue or self.queue, self.base_data, ary,
                     device_offset=self.offset,
                     is_blocking=not async_)
+            if not async_:  # not already waited for
```

Comment:
I would actually argue that this event should be added unconditionally, independent of `async_`, for consistency. Compute kernels are also "async" in that sense, and their events end up in the list.

Reply:
While we're at it--could you add a sentence to the docstring documenting this behavior?

Reply:
That's actually a bug - as noted in the comment, it's meant to be an "only add events we haven't already waited for" optimization, but that's "if async_", i.e. it's the wrong way round.

Which suggests this optimization isn't worth the risk of creating bugs, so the event is now unconditionally added.